app-cpp/appPtr.hpp

```diff
@@ -0,0 +1,394 @@
+                {
+                    delete mData;
+                    if (mControl->GetRefCount().second <= 0)
+                    {
+                        delete mControl;
+                    }
+                }
+            }
+            mData = nullptr;
+            mControl = nullptr;
+        }
+
+        // RO5 Methods
+        // Move Ctor
+        shared_ptr(shared_ptr<T>&& ioMove) : 
+            mData{ioMove.mData},
+            mControl{ioMove.mControl}
+        {
+            ioMove.mData = nullptr;
+            ioMove.mControl = nullptr;
+        }
+
+        // Move Operand
+        shared_ptr& operator=(shared_ptr<T>&& ioMove)
+        {
+            // Moving to self should be a NOP
+            if (this != &ioMove)
+            {
+                std::swap(mData, ioMove.mData);
+                std::swap(mControl, ioMove.mControl);
+            }
+            return *this;
+        }
+
+        // Copy Ctor
+        shared_ptr(const shared_ptr &inCopy)
+        {
+            mData = inCopy.mData;
+            mControl = inCopy.mControl;
+            mControl->IncrementStrong();
+        }
+
+        // Copy Operand
+        shared_ptr& operator=(const shared_ptr &inCopy)
+        {
+            // Copying to self should be a NOP
+            if (this != &inCopy)
+            {
+                mData = inCopy.mData;
+                mControl = inCopy.mControl;
+                mControl->IncrementStrong();
+            }
+            return *this;
+        }
+
+        T* operator->()
+        {
+            return get();
+        }
+
+        T& operator*()
+        {
+            return *get();
+        }
+
+        operator bool() const
+        {
+            return (mData != nullptr);
+        }
+
+        T* get()
+        {
+            return mData;
+        }
+
+    private:
+        class ControlBlock
+        {
+        public:
+            ControlBlock(T* inData) :  
+                 mData{inData}
+            {
+                // Do nothing
+            }
+
+            ControlBlock(shared_ptr<T>* inPtr) :
+                 mData{inPtr->mData}
+            {
+                // Do nothing
+            }
+
+            // RO5 Methods added
+            ~ControlBlock() = default;
+
+            T* get()
+            {
+                return mData;
+            }
+
+            std::pair<std::size_t, std::size_t> GetRefCount()
+            {
+                std::lock_guard<std::mutex> lock{mMutex};
+                
+                std::pair<std::size_t, std::size_t> refCounts{mStrongCount, mWeakCount};
+                return refCounts;
+            }
+
+            std::pair<std::size_t, std::size_t> IncrementStrong()
+            {
+                std::lock_guard<std::mutex> lock{mMutex};
+                std::pair<std::size_t, std::size_t> refCounts{++mStrongCount, mWeakCount};
+                return refCounts;
+            }
+
+            std::pair<std::size_t, std::size_t> DecrementStrong()
```

Comment:
I use pair since there are only 2 control elements for shared_ptr and weak_ptr. As I will never have to add more control types and the strong and weak counts are closely related, I opted for std::pair. Tuple would have a similar effect, however, it could incorrectly suggest that the control elements might change in the future.